alibi_detect/cd/tabular.py

```diff
@@ -0,0 +1,130 @@
+            preprocess_X_ref=preprocess_X_ref,
+            update_X_ref=update_X_ref,
+            preprocess_fn=preprocess_fn,
+            preprocess_kwargs=preprocess_kwargs,
+            correction=correction,
+            n_features=n_features,
+            n_infer=n_infer,
+            input_shape=input_shape,
+            data_type=data_type
+        )
+        if isinstance(categories_per_feature, dict):
+            if None not in list(categories_per_feature.values()):
+                # convert from Dict[int, int] to Dict[int, List[int]]
+                self.categories_per_feature = {f: list(np.arange(v)) for f, v in categories_per_feature.items()}
+            else:  # infer number of possible categories for each categorical feature from reference data
+                X_flat = self.X_ref.reshape(self.X_ref.shape[0], -1)
+                self.categories_per_feature = {f: list(np.unique(X_flat[:, f]))
+                                               for f in categories_per_feature.keys()}
+
+            if update_X_ref is None and preprocess_X_ref:
+                # already infer categories and frequencies for reference data
+                self.X_ref_count = {f: [(self.X_ref[:, f] == v).sum() for v in vals] for f, vals in
+                                    self.categories_per_feature.items()}
+            else:
+                self.X_ref_count = None
+        else:  # no categorical features assumed present
+            self.categories_per_feature, self.X_ref_count = {}, None
+
+    def feature_score(self, X_ref: np.ndarray, X: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Compute K-S or Chi-Squared test statistics and p-values per feature.
+
+        Parameters
+        ----------
+        X_ref
+            Reference instances to compare distribution with.
+        X
+            Batch of instances.
+
+        Returns
+        -------
+        Feature level p-values and K-S or Chi-Squared statistics.
+        """
+        X = X.reshape(X.shape[0], -1)
+        X_ref = X_ref.reshape(X_ref.shape[0], -1)
+        if self.categories_per_feature:
+            X_count = {f: [(X[:, f] == v).sum() for v in vals] for f, vals in
+                       self.categories_per_feature.items()}
+            if not self.X_ref_count:  # compute categorical frequency counts for each feature
+                X_ref_count = {f: [(X_ref[:, f] == v).sum() for v in vals] for f, vals in
+                               self.categories_per_feature.items()}
+            else:
+                X_ref_count = self.X_ref_count
+        p_val = np.zeros(self.n_features, dtype=np.float32)
+        dist = np.zeros_like(p_val)
+        for f in range(self.n_features):
+            if f in list(self.categories_per_feature.keys()):
+                dist[f], p_val[f] = chisquare(X_ref_count[f], f_exp=X_count[f])
+            else:
+                dist[f], p_val[f] = ks_2samp(X_ref[:, f], X[:, f], alternative='two-sided', mode='asymp')
```

Comment:
Does it make sense to "bake in" these options here as opposed to `KSDrift` which exposes these to the user?

Reply:
Yea I think I agree, maybe some disclaimer in the docs about how to use this feature.

Reply:
Agreed.